### PR TITLE
Add Connection Tracking Service

### PR DESCRIPTION
**Pull request by koolzz**
Signed-off-by: Nick Yurchenko <koolzz@fb.com>

## Summary

Adds a new magma service connection_tracker
 - Used libmnl for monitoring netlink connection tracking events.
 - Used libtins for crafting internal packets to trigger the ipfix export.

**_This would ideally require an integ test.. leaving that up to future PR_**

## Test Plan

Run locally and watch tcpdump

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/protos/mconfig/mconfigs.proto
Comment (ardzoht) [inline lte/protos/mconfig/mconfigs.proto:359]: make sure to generate proto file from this change

Comment (ardzoht) [reply lte/protos/mconfig/mconfigs.proto:359]: I'd also recommend adding the default entry for connectiond service on `gateway.mconfig`

File: lte/gateway/deploy/roles/magma/files/systemd/magma_connectiond.service
Comment (ardzoht) [inline lte/gateway/deploy/roles/magma/files/systemd/magma_connectiond.service]: Update the description

File: lte/gateway/c/connection_tracker/PacketGenerator.h
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/PacketGenerator.h:27]: since this will live under `/lte/gateway/` let's use the same namespaces for the services we already have, this would be 
```
namespace magma {
namespace lte {
} // namespace lte 
} // namespace magma
```

the same for all header / cpp files

File: lte/gateway/c/connection_tracker/PacketGenerator.h
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/PacketGenerator.h]: can we pass these by const reference?

`const std::string& iface_name`...

File: lte/gateway/c/connection_tracker/PacketGenerator.h
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/PacketGenerator.h:40]: I think `struct` symbol in this case since it's C++, it's optional, I'll leave it up to you if you want to be explicit but it's not required

File: lte/gateway/c/connection_tracker/PacketGenerator.cpp
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/PacketGenerator.cpp]: let's update these to be passed as const reference

File: lte/gateway/c/connection_tracker/EventTracker.h
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/EventTracker.h]: there's repeated header files being included here, let's remove duplicates

File: lte/gateway/c/connection_tracker/EventTracker.h
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/EventTracker.h]: I'd recommend to only include files that the header will need, as minimum as possible, and then include all necessary header files on the source .cpp file, let's clean this up with only the necessary include files for the header file and move all of the other to the EventTracker.cpp file

File: lte/gateway/c/connection_tracker/EventTracker.cpp
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/EventTracker.cpp:88]: Can we move these helper functions to a separate place? Maybe `EventTrackerHelpers` and then include that here, ideally this source file should only contain the definition of the `EventTracker` member class functions

File: lte/gateway/c/connection_tracker/CMakeLists.txt
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/CMakeLists.txt]: is there plans to add event logging to these? not sure why this was added since I couldn't find usages for this

File: lte/gateway/c/connection_tracker/CMakeLists.txt
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/CMakeLists.txt]: sessionD => connectiond

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/c/connection_tracker/EventTracker.cpp
Comment (koolzz) [reply lte/gateway/c/connection_tracker/EventTracker.cpp:88]: these are local static helper functions, isn't the best practice to just keep in the file that uses it here? I'm open to moving them though, should I just stil them into EventTrackerHelpers.h?

Comment (ardzoht) [reply lte/gateway/c/connection_tracker/EventTracker.cpp:88]: Ok I'm cool keeping them if they are bound to EventTracker usage, I'd recommend in this case moving upwards the definitions of the EventTracker member functions, and then the helpers.

File: lte/gateway/c/connection_tracker/EventTracker.cpp
Comment (pshelar) [inline lte/gateway/c/connection_tracker/EventTracker.cpp:267]: Can we use this zoneid from pipelieD to filter out unnecessary conntrack events. the zone id can be passed by .yml file.

File: lte/gateway/c/connection_tracker/EventTracker.cpp
Comment (pshelar) [inline lte/gateway/c/connection_tracker/EventTracker.cpp:79]: netlink can loose data in case of heavy load. can you investigate how to avoid it and detect it if it is un avoidable.
We need to handle the case by polling connections from kernel table. It can be TODO for now. But we need mechanism to identify that we lost events from conntrack.

**Review by pshelar — CHANGES_REQUESTED** [review]

File: lte/gateway/c/connection_tracker/EventTracker.cpp
Comment (koolzz) [reply lte/gateway/c/connection_tracker/EventTracker.cpp:79]: Interesting, okay I'll research and have a task for it

File: lte/gateway/c/connection_tracker/EventTracker.cpp
Comment (koolzz) [reply lte/gateway/c/connection_tracker/EventTracker.cpp:267]: Yeah its the plan, currently tb[CTA_MARK] is unset for some reason, I can't figure out why :|

Comment (koolzz) [reply lte/gateway/c/connection_tracker/EventTracker.cpp:267]: As I can validate basic functionality without this, I'd preffer to leave it as a todo

File: lte/gateway/c/connection_tracker/CMakeLists.txt
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/CMakeLists.txt]: nit: I don't think you need generating protos here, so we can remove these

File: lte/gateway/c/connection_tracker/CMakeLists.txt
Comment (ardzoht) [inline lte/gateway/c/connection_tracker/CMakeLists.txt:50]: remove async_grpc as it's unused on the library

**Review by ardzoht — APPROVED** [review]

**Comment by koolzz** [comment]
nothing to see here folks, no need for everyone to review

**Review by pshelar — APPROVED** [review]
Looks good to me.

File: lte/gateway/c/connection_tracker/CMakeLists.txt
Comment (koolzz) [reply lte/gateway/c/connection_tracker/CMakeLists.txt]: needed for mconfig

**Comment by koolzz** [comment]
*don't merge yet*
trying to figure out why including libs to https://github.com/magma/magma/pull/2318/commits/567f4d784ac0e3b8d8555e57269eaed0229c9dbc doesn't install them on provision, what is this new preburn thing we have

**Review by xjtian — APPROVED** [review]
Rubberstamping

**Comment by ssanadhya** [comment]
@koolzz could you please run S1ap integration tests on this change? Jenkins check is failing on traffic tests.

**Pull request merged**
